### ISSUE TITLE
SD TTNN: TtCrossAttention [Pending CI]

### ISSUE THREAD

**Issue by boris-drazic** [tenstorrent-metal/tt-metal#4792]
Implement subgraph `TtCrossAttention` and make two tests: one for 256x256 image size and one for 512x512 image size.
Have one test file named something like `test_cross_attention.py` and inside it two tests `test_cross_attention_256x256` and `test_cross_attention_512x512`. Tests should be parametrized with following input shapes:

- [x] Shape 1:  **256x256**
    - [x] (Case 1) hidden_states: [1, 2, 1024, 320], encoder_hidden_states: None
    - [x] (Case 2) hidden_states: [1, 2, 1024, 320], encoder_hidden_states: [1, 2, 77, 768]
    - [x] (Case 3) hidden_states: [1, 2, 256, 640], encoder_hidden_states: None
    - [x] (Case 4) hidden_states: [1, 2, 256, 640], encoder_hidden_states: [1, 2, 77, 768]
    - [x] (Case 5) hidden_states: [1, 2, 64, 1280], encoder_hidden_states: None
    - [x] (Case 6) hidden_states: [1, 2, 64, 1280], encoder_hidden_states: [1, 2, 77, 768]
    - [x] (Case 7) hidden_states: [1, 2, 16, 1280], encoder_hidden_states: None
    - [x] (Case 8) hidden_states: [1, 2, 16, 1280], encoder_hidden_states: [1, 2, 77, 768]

- [x] Shape 2:  **512x512**
    - [x] (Case 1) hidden_states: [1, 2, 4096, 320], encoder_hidden_states: None
    - [x] (Case 2) hidden_states: [1, 2, 4096, 320], encoder_hidden_states: [1, 2, 77, 768]
    - [x] (Case 3) hidden_states: [1, 2, 1024, 640], encoder_hidden_states: None
    - [x] (Case 4) hidden_states: [1, 2, 1024, 640], encoder_hidden_states: [1, 2, 77, 768]
    - [x] (Case 5) hidden_states: [1, 2, 256, 1280], encoder_hidden_states: None
    - [x] (Case 6) hidden_states: [1, 2, 256, 1280], encoder_hidden_states: [1, 2, 77, 768]
    - [x] (Case 7) hidden_states: [1, 2, 64, 1280], encoder_hidden_states: None
    - [x] (Case 8) hidden_states: [1, 2, 64, 1280], encoder_hidden_states: [1, 2, 77, 768]

For all the 16 cases mentioned above,
- attention_mask: None
- cross_attention_kwargs: {}

**Comment by vigneshkeerthivasanx** [comment]
This commit includes the implementation of cross-attention submodule for Stable diffusion model.

The tests contains the cases, where both encoder_hidden_states is not None.
For the following cases we are getting a PCC < 0.22 and for other tests we are getting PCC >= 0.99.

Shape 1: 256x256
hidden_states: [1, 2, 64, 1280], encoder_hidden_states: [1, 2, 77, 768]
PCC : 0.14407504034331337

Shape 2: 512x512
hidden_states: [1, 2, 256, 1280], encoder_hidden_states: [1, 2, 77, 768]
PCC: 0.22450454818184884

hidden_states: [1, 2, 64, 1280], encoder_hidden_states: [1, 2, 77, 768]
PCC: 0.2231018312226474

For the cases where encoder_hidden_states = None. We are getting  shape mismatches in few places. We are working on it.

Commit: [Link](https://github.com/tenstorrent-metal/tt-metal/compare/vignesh/stable_diffusion_tt_cross_attention)

**Comment by vigneshkeerthivasanx** [comment]
Implementing TtCrossAttention using ttnn for the Stable diffusion model is Completed.
Corresponding PR : [Link](https://github.com/tenstorrent-metal/tt-metal/pull/4927)

For the following cases we are getting a PCC < 0.22 and for other tests we are getting PCC >= 0.99.

Shape 1: 256x256

- hidden_states: [1, 2, 64, 1280], encoder_hidden_states: [1, 2, 77, 768]
  PCC : 0.14407504034331337

Shape 2: 512x512

- hidden_states: [1, 2, 256, 1280], encoder_hidden_states: [1, 2, 77, 768]
  PCC: 0.22450454818184884

- hidden_states: [1, 2, 64, 1280], encoder_hidden_states: [1, 2, 77, 768]
  PCC: 0.2231018312226474

**Comment by vigneshkeerthivasanx** [comment]
The updated PR improves the PCC, for all the test cases PCC = 0.99.
Corresponding PR : #4927  

Two tensors, values and attention_probs, both in TILE_LAYOUT, are utilised in ttnn.matmul. Shape of the two tensors for a sample case is given below.


```
   attention_probs: [1, 16, 1024, 80 + 16]
   value: [1, 16, 77 + 19, 80 + 16]
```


Despite both the tensors being in TILE_LAYOUT and padded automatically, we encountered a runtime error in ttnn.matmul(attention_probs, value)

`RuntimeError: The width of the first tensor must be equal to the height of the second tensor.`

As a workaround we have to manually pad the above tensors using ttnn.pad to make ttnn.matmul work.
Padding the attention_probs' 4th dimension to align with the shape of values' 3rd dimension.

```
         padding_needed = value.shape[-2] - attention_probs.shape[-1]
         attention_probs = ttnn.pad(attention_probs,((0,0),(0,padding_needed)), 0)
```

In the above case we were facing PCC drop. 0.2<PCC<0.8.

Padding the values' 3rd dimension to align with the shape of attention_probs' 4th dimension.
     

```
        padding_needed = attention_probs.shape[-1] - value.shape[-2]
        value = ttnn.pad(value, ((0, padding_needed), (0, 0)), 0)
```

In this case we were able to get PCC = 0.99.
